### PR TITLE
Fixes for ERC20 bridge deployment

### PR DESCRIPTION
**Pull request by Maximkaaa**
## Issue ticket

Issue ticket link: https://infinityswap.atlassian.net/browse/EPROD-1034

### Code conventions

- [ ] I have performed a self-review of my code
- [ ] Every new function is documented
- [ ] Object names are auto explicative
- [ ] The PR follows the [project conventions](https://infinityswap.atlassian.net/wiki/spaces/CPROD/pages/23330839/Conventions)
- [ ] The code follows the [style guidelines of the project](https://infinityswap.atlassian.net/wiki/spaces/CPROD/pages/24444929/Rust+Conventions)
- [ ] The code follows the [project security best practices](https://infinityswap.atlassian.net/wiki/spaces/CPROD/pages/35225620/Security+Considerations)

### Security 

- [ ] The PR does not break APIs backward compatibility
- [ ] The PR does not break the stable storage backward compatibility

### Testing 

- [ ] Every function is properly unit tested
- [ ] I have added integration tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] IC endpoints are always tested through the `canister_call!` macro

### CODE REVIEW

**Review by ufoscout — CHANGES_REQUESTED** [review]
It's ok but there are some `println!` usages to be removed

**Review by veeso — APPROVED** [review]

File: just/deploy.just [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+            coinbase = \"0x0000000000000000000000000000000000000000\"; \
+            log_settings = opt record { log_filter = opt \"info\"; in_memory_records = opt 10000; enable_console = true } \
+        })"
+
+[group('deploy')]
+configure_signature:
+    dfx canister call signature_verification add_access "(principal \"$(dfx canister id evm_testnet)\")"
+
+[group('deploy')]
+deploy_evm:
```
Comment (ufoscout) [inline just/deploy.just:23]: this should be:
```
[group('deploy')]
deploy_evm: deploy_signature deploy_evm_canister configure_signature
```
The advantage of this is that it enables static checks on the recipe names

File: src/bridge-deployer/src/bridge_deployer.rs [hunk 2]
```diff
@@ -55,7 +55,7 @@ impl BridgeDeployer {
 
         let canister_id = self.client.client().canister_id;
         let management_canister = ManagementCanister::create(&self.agent);
-        let arg = config.init_raw_arg(network)?;
+        let arg = config.init_raw_arg(self.agent.get_principal().unwrap(), network)?;
```
Comment (ufoscout) [inline src/bridge-deployer/src/bridge_deployer.rs:58]: Replace `unwrap()` with `?`

File: src/bridge-deployer/src/cli.rs [hunk 3]
```diff
@@ -72,13 +80,18 @@ pub struct Cli {
...
     pub async fn run() -> anyhow::Result<()> {
+        let _ = dotenv::dotenv();
         let cli = Cli::parse();
 
         // Initialize tracing with the appropriate log level based on the verbosity setting.
         cli.init_tracing();
+        let identity = cli.init_identity()?;
+        info!(
+            "Using dfx identity with principal: {}",
+            identity.sender().unwrap()
```
Comment (ufoscout) [inline src/bridge-deployer/src/cli.rs:91]: remove `unwrap`

File: src/bridge-deployer/src/commands/deploy.rs [hunk 4]
```diff
@@ -124,16 +129,45 @@ impl DeployCommands {
...
-            canister_type = self.bridge_type.kind(),
-            canister_id = deployer.bridge_principal(),
+            "Canister {} deployed successfully.",
+            self.bridge_type.kind(),
         );
+        println!("Bridge canister principal: {}", canister_id);
+        println!();
+        println!("Wrapped side BFT bridge: {}", bft_bridge);
+        println!("Wrapped side FeeCharge: {}", fee_charge);
+        println!("Wrapped side bridge address: {}", minter_address);
```
Comment (ufoscout) [inline src/bridge-deployer/src/commands/deploy.rs:157]: should we log all of this data instead?

File: src/bridge-deployer/src/cli.rs [hunk 5]
```diff
@@ -166,4 +180,40 @@ impl Cli {
     fn filter_none(_metadata: &tracing::Metadata) -> bool {
         true
     }
+
+    fn init_identity(&self) -> anyhow::Result<GenericIdentity> {
```
Comment (ufoscout) [inline src/bridge-deployer/src/cli.rs:184]: Add fn description

File: src/bridge-deployer/src/commands/deploy.rs [hunk 6]
```diff
@@ -179,4 +216,29 @@ impl DeployCommands {
 
         Ok(())
     }
+
+    fn get_wallet_canister(&self, network: EvmNetwork) -> anyhow::Result<Principal> {
```
Comment (ufoscout) [inline src/bridge-deployer/src/commands/deploy.rs:220]: Add fn description

File: src/bridge-deployer/src/commands/mod.rs [hunk 7]
```diff
@@ -163,14 +172,44 @@ impl Bridge {
         Ok(arg)
     }
 
-    /// Returns if the bridge is wrapped side or not
-    pub fn is_wrapped_side(&self) -> bool {
+    pub async fn finalize(
```
Comment (ufoscout) [inline src/bridge-deployer/src/commands/mod.rs:175]: Add fn description

File: src/bridge-deployer/src/commands/mod.rs [hunk 8]
```diff
@@ -232,47 +272,64 @@ pub struct BFTArgs {
...
         pk: H256,
         agent: &Agent,
+        is_wrapped_side: bool,
     ) -> anyhow::Result<BftDeployedContracts> {
         info!("Deploying BFT bridge");
 
         let contract_deployer = SolidityContractDeployer::new(network, pk);
 
-        let expected_nonce = contract_deployer.get_nonce().await? + 3;
+        let nonce_increment = match is_wrapped_side {
```
Comment (ufoscout) [inline src/bridge-deployer/src/commands/mod.rs:297]: Add a short description of why this is needed and what is the underlying logic

Comment (ufoscout) [reply src/bridge-deployer/src/commands/mod.rs:297]: This also seems fragile, if in the future we add another step before this one then the nonce will be different. Can it be replaced with an atomic counter passed as a parameter to the functions that need it?

**Review by ufoscout — CHANGES_REQUESTED** [review]

File: src/bridge-deployer/src/bridge_deployer.rs (hunk 2 above)
Comment (Maximkaaa) [reply src/bridge-deployer/src/bridge_deployer.rs:58]: This function returns `String` error which cannot be cleanly converted to `anyhow::Error`. But this code should never fail with valid identity, so I just replaced `unwrap` with `expect`.

File: src/bridge-deployer/src/cli.rs (hunk 3 above)
Comment (Maximkaaa) [reply src/bridge-deployer/src/cli.rs:91]: This function returns `String` error which cannot be cleanly converted to `anyhow::Error`. But this code should never fail with valid identity, so I just replaced `unwrap` with `expect`.

File: src/bridge-deployer/src/commands/mod.rs (hunk 8 above)
Comment (Maximkaaa) [reply src/bridge-deployer/src/commands/mod.rs:298]: There is comment about that 2 lines below explaining this logic. I agree that this looks like a hack, but I don't know any other way to predict the nonce. We need to calculate the nonce before we deploy BftBridge contract, so before any operations are actually done.

File: src/bridge-deployer/src/commands/deploy.rs (hunk 4 above)
Comment (veeso) [reply src/bridge-deployer/src/commands/deploy.rs:157]: Don't print the empty line please, since it's annoying for the integration tests.

**Pull request merged**